cmd/client-s3.go

```diff
@@ -315,15 +315,78 @@ func (c *S3Client) RemoveNotificationConfig(arn string) *probe.Error {
 	}
 	accountArn := minio.NewArn(fields[1], fields[2], fields[3], fields[4], fields[5])
 
-	switch fields[2] {
-	case "sns":
-		mb.RemoveTopicByArn(accountArn)
-	case "sqs":
-		mb.RemoveQueueByArn(accountArn)
-	case "lambda":
-		mb.RemoveLambdaByArn(accountArn)
-	default:
-		return errInvalidArgument().Trace(fields[2])
+	// if we are passed filters for either events, suffix or prefix, then only delete the single event that matches
+	// the arguments
+	if event != "" || suffix != "" || prefix != "" {
+		// Translate events to type events for comparison
+		events := strings.Split(event, ",")
+		var eventsTyped []minio.NotificationEventType
+		for _, e := range events {
+			switch e {
+			case "put":
+				eventsTyped = append(eventsTyped, minio.ObjectCreatedAll)
+			case "delete":
+				eventsTyped = append(eventsTyped, minio.ObjectRemovedAll)
+			case "get":
+				eventsTyped = append(eventsTyped, minio.ObjectAccessedAll)
+			default:
+				return errInvalidArgument().Trace(events...)
+			}
+		}
+		// based on the arn type, we'll look for the event in the corresponding sublist and delete it if there's a match
+		switch fields[2] {
+		case "sns":
+			removeIndex := -1
+			for i, v := range mb.TopicConfigs {
+				// if it matches events and filters, mark the index for deletion
+				if EqualNotificationConfig(v.NotificationConfig, arn, eventsTyped, prefix, suffix) {
+					removeIndex = i
+					break // since we have at most one matching config
+				}
+			}
+			if removeIndex >= 0 {
+				mb.TopicConfigs = append(mb.TopicConfigs[:removeIndex], mb.TopicConfigs[removeIndex+1:]...)
+			}
+		case "sqs":
+			removeIndex := -1
+			for i, v := range mb.QueueConfigs {
+				// if it matches events and filters, mark the index for deletion
+				if EqualNotificationConfig(v.NotificationConfig, arn, eventsTyped, prefix, suffix) {
+					removeIndex = i
+					break // since we have at most one matching config
+				}
+			}
+			if removeIndex >= 0 {
+				mb.QueueConfigs = append(mb.QueueConfigs[:removeIndex], mb.QueueConfigs[removeIndex+1:]...)
+			}
+		case "lambda":
+			removeIndex := -1
+			for i, v := range mb.LambdaConfigs {
+				// if it matches events and filters, mark the index for deletion
+				if EqualNotificationConfig(v.NotificationConfig, arn, eventsTyped, prefix, suffix) {
+					removeIndex = i
+					break // since we have at most one matching config
+				}
+			}
+			if removeIndex >= 0 {
+				mb.LambdaConfigs = append(mb.LambdaConfigs[:removeIndex], mb.LambdaConfigs[removeIndex+1:]...)
```

Comment:
Ideally, we should add this code in `minio-go` such that it provides `RemoveTopicByPrefixSuffix`. It is better to create a set.NewStringSet() and use that to instead find commonalities - you can avoid complex sorting etc.

Reply:
Should I move this remove to `minio-go` and use the stringsets ?

Reply:
Ideally yes @dvaldivia

Reply:
ready and passing @harshavardhana https://github.com/minio/minio-go/pull/1262